### PR TITLE
Fix bug where interpolations were unnecessarily resolved during merge.

### LINKED ISSUE #431: [Bug] omegaconf is doing interpolation during merges when it shouldnt
copied from hydra issue here:
https://github.com/facebookresearch/hydra/issues/1125

repro:

```python
from omegaconf import OmegaConf
from dataclasses import dataclass
from typing import List
@dataclass
class test_class:
    x: List[float] = "${optimization.lr}"

a = OmegaConf.structured(test_class())
b = OmegaConf.create({"x": [0.1]})
OmegaConf.merge(a,b)
```
Exception:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/private/home/abaevski/.conda/envs/fairseq-fp16-20200821/lib/python3.6/site-packages/omegaconf/omegaconf.py", line 321, in merge
    target.merge_with(*others[1:])
  File "/private/home/abaevski/.conda/envs/fairseq-fp16-20200821/lib/python3.6/site-packages/omegaconf/basecontainer.py", line 327, in merge_with
    self._format_and_raise(key=None, value=None, cause=e)
  File "/private/home/abaevski/.conda/envs/fairseq-fp16-20200821/lib/python3.6/site-packages/omegaconf/base.py", line 101, in _format_and_raise
    type_override=type_override,
  File "/private/home/abaevski/.conda/envs/fairseq-fp16-20200821/lib/python3.6/site-packages/omegaconf/_utils.py", line 675, in format_and_raise
    _raise(ex, cause)
  File "/private/home/abaevski/.conda/envs/fairseq-fp16-20200821/lib/python3.6/site-packages/omegaconf/_utils.py", line 591, in _raise
    raise ex  # set end OC_CAUSE=1 for full backtrace
omegaconf.errors.ConfigKeyError: str interpolation key 'optimization.lr' not found
	full_key:
	reference_type=Any
	object_type=test_class
```

if instead you set a to be DictConfig initialized from a dict, everything works
```
>>> a = DictConfig({'x': '${optimization.lr}'})
>>> OmegaConf.merge(a,b)
{'x': [0.1]}
```

### PR DESCRIPTION
**Pull request by pereman2**
Closes #431 .

DictConfig and ListConfig which were interpolations had to be "expanded" in order to fulfil the merge. If it wasn't, it detaches the node to find the interpolation value which in the example given in #431 doesn't exist and then fails. is_interpolation has to be checked before is_missing since is_missing detaches the node(which will resolve the interpolation) to find if that is missing.

### CODE REVIEW

File: tests/__init__.py [hunk 1]
```diff
@@ -154,3 +154,13 @@ class SubscriptedList:
 @dataclass
 class SubscriptedDict:
     dict: Dict[str, int] = field(default_factory=lambda: {"foo": 4})
+
+
+@dataclass
+class InterpolationList:
+    list: List[float] = "${optimization.lr}"  # type: ignore
```
Comment (omry) [inline tests/__init__.py:161]: Check the docs for interpolation with Structured Configs.
you can do something like:
```
list: List[float] = II("optimization.lr")
```
And this would not need the type: ignore.

File: omegaconf/basecontainer.py [hunk 2]
```diff
@@ -387,7 +387,7 @@ def _merge_with(
             if isinstance(self, DictConfig) and isinstance(other, DictConfig):
                 BaseContainer._map_merge(self, other)
             elif isinstance(self, ListConfig) and isinstance(other, ListConfig):
-                if self._is_none() or self._is_missing() or self._is_interpolation():
+                if self._is_none() or self._is_interpolation() or self._is_missing():
                     self.__dict__["_content"] = []
                 else:
                     self.__dict__["_content"].clear()
```
Comment (omry) [inline omegaconf/basecontainer.py:393]: This bit puzzles me. the two branches are pretty similar.
Can you try to see if you can just use one of the branches always?

File: tests/test_merge.py [hunk 3]
```diff
@@ -484,3 +486,23 @@ def test_merge_allow_objects() -> None:
...
+            {"list": [0.1]},
+        ),
+        (
+            OmegaConf.structured(InterpolationDict),
+            OmegaConf.create({"dict": {"a": 4}}),
+            {"dict": {"a": 4}},
+        ),
+    ],
+)
+def test_merge_with_interpolation(dst: Any, other: Any, expected: Any) -> None:
```
Comment (omry) [inline tests/test_merge.py:506]: Add a test for merging with itself.

Should be something like:
```
            OmegaConf.structured(InterpolationDict),
            OmegaConf.structured(InterpolationDict),
            {"dict": {"a": "${optimization.lr}"}},
```
I want to see that the interpolation is not resolved in the test.

**Review by omry — CHANGES_REQUESTED** [review]
Looking good!
Got a few requests.

**Comment by omry** [comment]
Also, can you tell me what is missing from this PR? :)

**Comment by pereman2** [comment]
> Also, can you tell me what is missing from this PR? :)

news fragment?

File: omegaconf/basecontainer.py (hunk 2 above)
Comment (pereman2) [reply omegaconf/basecontainer.py:393]: This block can be reduced to `self.__dict__["_content"] = []` without the if-else and works the same

File: tests/test_merge.py (hunk 3 above)
Comment (pereman2) [reply tests/test_merge.py:506]: This fails, in this case in validate_merge

Comment (pereman2) [reply tests/test_merge.py:506]: image

**Comment by omry** [comment]
> > Also, can you tell me what is missing from this PR? :)
> 
> news fragment?

Yup :)

File: omegaconf/basecontainer.py (hunk 2 above)
Comment (omry) [reply omegaconf/basecontainer.py:393]: Cool, please do it.

File: tests/test_merge.py (hunk 3 above)
Comment (omry) [reply tests/test_merge.py:506]: Okay, this needs to pass too. no reason for it to fail.

File: omegaconf/basecontainer.py [hunk 4]
```diff
@@ -298,7 +301,7 @@ def expand(node: Container) -> None:
                 else:
                     node._set_value(type_)
 
-        if dest._is_missing():
+        if dest._is_interpolation() or dest._is_missing():
```
Comment (omry) [inline omegaconf/basecontainer.py:304]: Will this work?
```suggestion
        if dest._is_missing(throw_on_resolution_failure=False):
```

File: omegaconf/basecontainer.py [hunk 5]
```diff
@@ -575,9 +577,12 @@ def _item_eq(
     def _is_none(self) -> bool:
         return self.__dict__["_content"] is None
 
-    def _is_missing(self) -> bool:
+    def _is_missing(self, throw_on_resolution_failure: bool = True) -> bool:
```
Comment (omry) [inline omegaconf/basecontainer.py:580]: What if you change this function to not throw on resolution failure (without a parameter), does this cause any issues?

File: tests/test_merge.py [hunk 6]
```diff
@@ -484,3 +486,46 @@ def test_merge_allow_objects() -> None:
...
+            "dict",
+            True,
+        ),
+        (
+            OmegaConf.structured(InterpolationList),
+            OmegaConf.structured(InterpolationList),
+            None,
+            "list",
+            True,
+        ),
```
Comment (omry) [inline tests/test_merge.py:521]: can you name tests?
pytest.param(..., id="test_name")

**Review by omry — COMMENTED** [review]
Nice. a few questions inline.

File: omegaconf/basecontainer.py (hunk 4 above)
Comment (pereman2) [reply omegaconf/basecontainer.py:304]: No, it doesn't.

File: omegaconf/basecontainer.py (hunk 5 above)
Comment (pereman2) [reply omegaconf/basecontainer.py:580]: It doesn't cause any issue, It could be changed to that.

File: omegaconf/basecontainer.py (hunk 4 above)
Comment (omry) [reply omegaconf/basecontainer.py:304]: Oh, of course.

File: omegaconf/basecontainer.py [hunk 7]
```diff
@@ -281,6 +281,9 @@ def _map_merge(dest: "BaseContainer", src: "BaseContainer") -> None:
         assert isinstance(src, DictConfig)
         src_type = src._metadata.object_type
 
+        if src._is_interpolation():
+            dest._set_value(src._value())
+            return
```
Comment (omry) [inline omegaconf/basecontainer.py:286]: Please explain this one.

File: omegaconf/dictconfig.py [hunk 8]
```diff
@@ -445,7 +445,7 @@ def pop(self, key: Union[str, Enum], default: Any = DEFAULT_VALUE_MARKER) -> Any
             self._format_and_raise(key=key, value=None, cause=e)
 
     def keys(self) -> Any:
-        if self._is_missing() or self._is_interpolation() or self._is_none():
+        if self._is_interpolation() or self._is_missing() or self._is_none():
```
Comment (omry) [inline omegaconf/dictconfig.py:448]: not that _is_missing dies not throw, you can undo this change.

File: omegaconf/omegaconf.py [hunk 9]
```diff
@@ -563,6 +563,8 @@ def _get_obj_type(c: Any) -> Optional[Type[Any]]:
         elif isinstance(c, DictConfig):
             if c._is_none():
                 return None
+            elif c._is_interpolation():
+                return None
```
Comment (omry) [inline omegaconf/omegaconf.py:567]: Please explain.

File: tests/test_merge.py [hunk 10]
```diff
@@ -484,3 +486,50 @@ def test_merge_allow_objects() -> None:
...
+    ],
+)
+def test_merge_with_interpolation(
+    dst: Any, other: Any, expected: Any, node: Any, check_missing: bool
+) -> None:
+    res = OmegaConf.merge(dst, other)
+    if check_missing:
+        OmegaConf.is_missing(res, node)
+    else:
+        assert res == expected
```
Comment (omry) [inline tests/test_merge.py:535]: do one thing in each test.
you can add another test to test the behavior of _is_missing with interpolation.

File: omegaconf/basecontainer.py (hunk 7 above)
Comment (pereman2) [reply omegaconf/basecontainer.py:286]: Merging an interpolation to a container is basically keeping the interpolation in _dest_ therefore, dest's value should be src's value.
If I don't do this check here it will fail in `for key, src_value in src.items_ex(resolve=False):`(line 307). 
To sum up, it's the same as line 289 where `if src._is_missing(): dest._set_value("???") `

File: omegaconf/omegaconf.py (hunk 9 above)
Comment (pereman2) [reply omegaconf/omegaconf.py:567]: Now that is_missing doesn't throw on resolution failure it doesn't matter this line. I'll remove it.

File: tests/test_merge.py (hunk 10 above)
Comment (pereman2) [reply tests/test_merge.py:535]: I'm changing it to is_interpolation

File: omegaconf/basecontainer.py (hunk 7 above)
Comment (omry) [reply omegaconf/basecontainer.py:286]: Ok. See that there is a comment explaining it in 288.
Add a similar comment.

**Review by omry — APPROVED** [review]
Great.
Merging, can you add another PR cherry-picking this into the 2.0_branch?

**Comment by pereman2** [comment]
> Great.
> Merging, can you add another PR cherry-picking this into the 2.0_branch?

Sure
